### PR TITLE
Add event for reporting updates (assets, props file, targets file writes) to PackageReference projects

### LINKED ISSUE #9782: NuGet restore in VS should report progress to allow the project-system to better control design time build scheduling.
CPS side: [1146684](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1146684)

Right now the BulkFileOperation API does not allow overlapping intervals. 

Internal issue: [1418717](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1418717/)

Design: https://github.com/NuGet/Home/blob/dev/proposed/2021/ProjectUpdateEventsInVisualStudio.md

PR in progress: https://github.com/NuGet/NuGet.Client/compare/dev-nkolev92-restoreProgressEvents?expand=1

### PR DESCRIPTION
**Pull request by nkolev92**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9782

Regression? Last working version:

## Description

Designed and reasoning for this change captured in https://github.com/NuGet/Home/blob/dev/proposed/2021/ProjectUpdateEventsInVisualStudio.md. 

In here, the implementation covers the propagation of all the APIs, and the addition of the reporting in VS restore. 

Note that the events are not propagated in any of the commandline restore scenarios as they are unnecessary and there's no need for extra allocations.

Helpers for the code review: 

* IVsNuGetProgressReporter - API to report the solution start/end. Internal only.
* SolutionRestoreJob - Where all VS restores are triggered. This is where the events are wired in.
* IVsNuGetPackageReferenceProjectUpdateEvents - The new extensibility API.
* RestoreRunner - Where all the restores get triggered. In particular this is where the assets file writes happen.

This PR will have a follow up in https://github.com/NuGet/Home/issues/11432. 
The updates are not getting reported for PM UI/Installer API/PMC updates.
This makes the changes more manageable.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [x] Documentation PR or issue filled - https://github.com/NuGet/docs.microsoft.com-nuget/issues/2616
  - **OR**
  - [ ] N/A

### CODE REVIEW

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreRunnerTests.cs
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreRunnerTests.cs:796]: I changed some test code around and it caused these tests to fail. 

These simple changes fix them.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/IVsNuGetProgressReporter.cs
Comment (jeffkl) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/IVsNuGetProgressReporter.cs:13]: Any reason to not have xml doc comments on this interface?  I see you added them to other ones

**Review by jeffkl — APPROVED** [review]

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/IVsNuGetProgressReporter.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/IVsNuGetProgressReporter.cs:13]: Honestly, no great reason :D 

I think it makes sense to add some docs here as well.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs]: When I debug your branch, I see null for `SolutionRestoreEventHandler`, value is not set. Just wondering why? For tests it explicitly set values.
image

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs]: Please refer to https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/event. 

Events will be null if there are no event handlers attached to it. 
Given that there are no consumers here, there are no event handlers when debugging and it'll be null.

It's basically *how it works*.

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs]: Who would be consumers? I don't recall the context. How it would be used in real life?

Comment (erdembayar) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs]: If I want to test this on my local, then how do I consume it?

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs]: The issue body has a link to the spec: https://github.com/NuGet/Home/blob/dev/proposed/2021/ProjectUpdateEventsInVisualStudio.md#summary. 

This is going to be a standard extensibility API, we're not introducing a new delivery mechanism . https://docs.microsoft.com/en-us/nuget/visual-studio-extensibility/nuget-api-in-visual-studio

**Review by erdembayar — APPROVED** [review]

**Comment by nkolev92** [comment]
This PR is still open because we do not want to merge it to 6.1, but to 6.2 directly.

**Comment by nkolev92** [comment]
Waiting on a review from Andy as per a recent offline discussion. 

This comment pretty much serves as a means to tell the bot to stay off my PR :D

File: src/NuGet.Clients/NuGet.VisualStudio/SolutionRestoreManager/IVsNuGetPackageReferenceProjectUpdateEvents.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio/SolutionRestoreManager/IVsNuGetPackageReferenceProjectUpdateEvents.cs]: I don't think this class needs to be specific to PackageReference projects. In fact, only dotnet/project-system knows whether a project is PackageReference or packages.config, because dotnet/project-system only works for SDK style projects, and SDK style projects only support PackageReference.

We know that Roslyn and Test Explorer both use `GetInstallPackagesAsync()`, and would be interested events letting them know when the package list might have changed, regardless of what restore style NuGet uses.

I'm not saying that we must make this interface work for packages.config in this PR, although not doing so would disappoint me, but I think we should drop `PackageReference` from the interface name to give it the opportunity to be extended to support packages.config in the future.

Our restore code already checks all PC projects in the solution, checks for any whose packages are missing from the solution packages folder, and then restores those packages. From a high level, it doesn't sound/feel like it should be difficult to also capture the list of projects that use those packages, and therefore include those project names in the `SolutionRestore[Started|Finished]` events, or trigger `ProjectUpdate[Started|Finished]` events for those same projects. Unlike PackageReference projects, PC projects' ProjectUpdate events would happen at the same time as SolutionRestore events in case of a restore.

I haven't yet looked though the rest of this PR, so I'm not sure if PC install/uninstall would be very different to PR package install/uninstall, but from the point of view of a VS extension author outside the NuGet team, the ProjectUpdateStarted and Finished events totally make sense for PC projects when they install, upgrade or uninstall packages though PMC, PM UI or any other VS feature like Roslyn's lightbulb code fixes.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:237]: I'm undecided what the best option is.  What do you think about using `EventSource` to raise an event instead, and have our VS code use `EventListener` to relay the event via the VS extensibility API?

From a NuGet team member point of view, the advantage would be that when customers report a problem and attach an ETL trace, we have more information that might help us debug some types of problems just a little bit more easily. However, I feel like we really should have ETW events for relevant restore information, even when it's a no-op (to be clear, this is out of scope for this PR, I'm just saying we need to eventually improve NuGet diagnosability). If the event raised for general restore information has a `noop` bool, then a separate ETW event for start/finish project update is redundant. On the other hand, if we add these generalised restore events with a noop property, then the `EventListener` could read the noop property value, and only raise the `IVs*` event when the value is false.

However, I also really, really don't want anyone, including our SDK customers, using our ETW events for business logic. In fact, EventSource and EventListener make code very decoupled in a way that is very difficult to test. Therefore, it's probably a bad idea to use it for our own business logic.

I think my gut reaction is to avoid public API changes, and using `EventSource` and `EventListener` would allow us to do that. But as I write down my thoughts in this comment, I'm pretty sure I've convinced myself that we should not use EventSource, even though it changes the public API. Which is a good thing for our SDK customers, since they'll be able to adopt the same feature. Even if we eventually replace PackageReference with something else in the (long distance) future, the concepts of a project will still exist, so supporting `IRestoreProgressReporter` doesn't seem unreasonable.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs]: What do you think about calling `GetInvocationList()`, and a for loop calling each one in a try-catch block, so that if any event subscribers have a but that thows an exception, we neither bubble the exception though our own code, and we don't prevent other subscribers from getting the events?

I have examples of this in my VS extensibility API ETW event PR, which I need to revive.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:244]: Doesn't doing this in `ExecuteAndCommitAsync` instead of `CommitAsync` mean that these events will not fire when a package is installed in PMUI, since it does the execute, show the preview window and then commit only if the customer accepts the preview changes?

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:54]: Should this method also have the obsolete attribute?  Or do we intend on "supporting" this API?

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:108]: At some point, should we consider making a request/args class, similar to how `NuGet.Commands` has a `RestoreArgs` class, so we no longer have to keep adding overloads when we add new features?

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs]: Some people might not like the syntax, but I think using Moq, you can write fewer lines of code and avoid the test class implementing the interface.

`var mock = Mock<IRestoreProgressReporter>()` will allow you to `mock.Validate(r => r.StartProjectUpdate(It.IsAny<string>(), It.IsAny<string>()), Times.One)` or `Times.Zero`. Instead of multiple dequeue lines, just create a new `Mock<T>` instance, as I don't think it provides the ability to clear its invocation list.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs
Comment (zivkan) [inline test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs]: Maybe I don't understand the test code well enough, but if projects are restored in parallel, doesn't that by definition make the order indeterministic?

**Review by zivkan — COMMENTED** [review]
> Waiting on a review from Andy as per a recent offline discussion.

sorry for being slow :(


I'm mostly done with my review, but ideally I'd like to re-read a thing or two. But I'm going to step away from the computer now for a while, and just in case I don't come back to this, I should post my comments now.

File: src/NuGet.Clients/NuGet.VisualStudio/SolutionRestoreManager/IVsNuGetPackageReferenceProjectUpdateEvents.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.VisualStudio/SolutionRestoreManager/IVsNuGetPackageReferenceProjectUpdateEvents.cs]: Considering that VS2022 discourages embedded interop types and we stopped shipping a props/targets file that enabled embedded interop types, perhaps we should stop adding `[ComImport]` to our interfaces. It's probably a bug that I didn't remove these `[ComImport]` attributes from all the interfaces when I stopped the package from using embed interop.

**Review by zivkan — COMMENTED** [review]
I'm a bit concerned there's a bug where events won't be raised when packages are installed via PM UI, and I think a test might have a timing bug, but mostly it looks ok to me. I think my other feedback are either questions or nitpicky things.

**Comment by nkolev92** [comment]
@zivkan 

> I'm a bit concerned there's a bug where events won't be raised when packages are installed via PM UI

I actually addressed that in the PR description. 

> This PR will have a follow up in NuGet/Home#11432.

I'll look at the rest of the feedback later.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs]: I'm checking OrderedEquals because of that.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:244]: I didn't read the PR comment. This is a known issue that will be followed up on.

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreRunner.cs:237]: This one was a trip :D 
But I did go through the same one originally. 

This is business logic, so I don't have an issue treating as such and being ok with the new API.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:108]: That's not a bad idea. 
I'd have a hard time deciding what belongs there. 

Everything? Only some APIs?

File: src/NuGet.Clients/NuGet.VisualStudio/SolutionRestoreManager/IVsNuGetPackageReferenceProjectUpdateEvents.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.VisualStudio/SolutionRestoreManager/IVsNuGetPackageReferenceProjectUpdateEvents.cs]: You're right, I don't think we need. 

I'll remove this.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:108]: not the cancellation token, but I think all the other parameters would belong in such a RestoreArgs class.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:54]: There are 40 tests calling this. 

We can obsolete and replace though. 

If you think we should do that, I can create a PR and make it call other methods instead.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:108]: Do you care if I do this in a future PR (right after this gets merged?)

I basically want to avoid others having to rereview this and this PR is already on the larger side.

Comment (zivkan) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:108]: sure, go for it.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:54]: It just seems weird to me that we mark the other method as obsolete, but not this one. We shouldn't "support" an API just because a test uses it, but a test using it and it being painful to change is an indication that it might be painful for our SDK customers to change as well. THerefore normally we should reconsider whether or not to obsolete the other API, but thi is the `NuGet.PackageManagement` project which I care less about.

File: test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs
Comment (nkolev92) [reply test/NuGet.Core.Tests/NuGet.PackageManagement.Test/BuildIntegration/DependencyGraphRestoreUtilityTests.cs]: I shifted it mock, but for the 3rd test, it's actually *way* longer than the original version.

File: src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:54]: The biggest reason why that method is obsolete is because it has unused parameters. 

In particular, the solution manager is not used, so I just wanted us to get rid of methods with unused parameters.

Comment (zivkan) [reply src/NuGet.Core/NuGet.PackageManagement/BuildIntegration/DependencyGraphRestoreUtility.cs:54]: Ok, cool. I didn't notice that. That is a great reason to obsolete a method!

**Review by zivkan — APPROVED** [review]

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/IVsNuGetProgressReporter.cs
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/IVsNuGetProgressReporter.cs:24]:
``` suggestion
        /// <param name="projects">Projects that were restored.</param>
```

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs
Comment (kartheekp-ms) [inline src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs:42]: I see this piece of code repeated several times in this PR. Can we abstract this logic into a utility method?

**Review by kartheekp-ms — COMMENTED** [review]
Magnificent work in this PR. I am slowly going through the changes and trying to understand the behavior.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/VsRestoreProgressEvents.cs:42]: Things are sort/kind of duplicated only once, given the parameters. 
This is code that we'll pretty much never change, so not too worried about the duplication.

File: src/NuGet.Clients/NuGet.SolutionRestoreManager/IVsNuGetProgressReporter.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.SolutionRestoreManager/IVsNuGetProgressReporter.cs:24]: I can make this change in the follow up work https://github.com/NuGet/Home/issues/11432. 

This PR has been open for a long time now, and I'd like to merge it so that our partners can try it out.

**Comment by nkolev92** [comment]
@kartheekp-ms 

Feel free to leave feedback where appropriate. 
Given that this PR has been open for quite a long time, I wanted to get it merged to allow our partners to try out the API. 

This change will go into 6.2, so I can address your feedback in a future PR. There'll be other PRs for this feature.

**Pull request merged**
